### PR TITLE
Feature/jenkins 38594 create pipeline begin

### PR DESCRIPTION
**Pull request by cliffmeyers**
# Description
- This one still needs a proper Dialog component from the JDL and some security applied. I opened this PR for now and am forking a new branch for JENKINS-38595 so I can make further progress.
- See [JENKINS-38594](https://issues.jenkins-ci.org/browse/JENKINS-38594).

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [x] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
  - Load up app, press "Create Pipeline": confirm classic job creation screen is launched in new tab.
  - Add ?blueCreate to the Blue Ocean URL, then click "Create Pipeline": confirm that a very crummy looking temp dialog overlays the main screen. Test that "CLOSE" and "back button" both navigate back to prior screen
  - Test that a "deep link" to the create pipeline route displays the dialog. It will overlay over a mostly empty screen, since there is no prior route.
- [x] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

**Comment by michaelneale** [comment]
nice good to see a good start on this!

File: blueocean-dashboard/src/main/js/PipelineRoutes.jsx [hunk 1]
```diff
@@ -82,9 +85,16 @@ function persistModalBackground() {
  * the background context.
  */
 function handleNavigationChangeToFromModal(prevState, nextState, replace, callback) {
-    if (nextState.params.runId && (prevState == null || !prevState.params.runId)) {
+    const enterRunDetails = nextState.params.runId && (prevState == null || !prevState.params.runId);
+    const leaveRunDetails = (prevState !== null && prevState.params.runId) && !nextState.params.runId;
+    const enterCreatePipeline = nextState.location.pathname.indexOf('/create-pipeline') !== -1 &&
+        (prevState == null || prevState.location.pathname.indexOf('/create-pipeline') === -1);
+    const leaveCreatePipeline = nextState.location.pathname.indexOf('/create-pipeline') === -1 &&
+        (prevState == null || prevState.location.pathname.indexOf('/create-pipeline') !== -1);
```
Comment (tfennelly) [inline blueocean-dashboard/src/main/js/PipelineRoutes.jsx]: @cliffmeyers hey Cliff, what are we trying to determine in the above code? It's a sea of `location.pathname.indexOf('/create-pipeline')` .... maybe we can compose this a bit more .... boil it down a bit and put it into a function(s) with some comments/javadoc etc

File: blueocean-dashboard/src/main/js/creation/CreatePipeline.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/**
+ * Created by cmeyers on 10/17/16.
+ */
+import React, { PropTypes } from 'react';
+
+function DialogPlaceholder(props) {
+    function closeHandler() {
```
Comment (scherler) [inline blueocean-dashboard/src/main/js/creation/CreatePipeline.js]: Why not:
```
const closeHandler = (props) => props.onClose || {};
```

File: blueocean-dashboard/src/main/js/PipelineRoutes.jsx (hunk 1 above)
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/js/PipelineRoutes.jsx]: @tfennelly made changes in 544c866. Let me know if that helps?

**Comment by cliffmeyers** [comment]
@michaelneale i added logic to show/hide pipeline similar to login button - actually refactored that code a bit so we have some more explicit utils in the "Security" class I created in core-js. The last remaining thing in this PR is to integrate a Dialog component. Should we just merge this sucker in with my ugly temp dialog or leave it on-hold until the Dialog is ready? I am still focused on the workflow stuff ATM.

File: blueocean-web/src/main/js/main.jsx [hunk 3]
```diff
@@ -12,13 +12,13 @@ import Config from './config';
...
-import { AppConfig, UrlConfig, Utils} from '@jenkins-cd/blueocean-core-js';
+import { AppConfig, Security, UrlConfig, Utils} from '@jenkins-cd/blueocean-core-js';
 
 let config; // Holder for various app-wide state
 
 function loginOrLogout() {
-    if (AppConfig.getLoginUrl()) {
-        if (AppConfig.getInitialUser() === "anonymous") {
+    if (Security.isSecurityEnabled()) {
+        if (Security.isAnonymousUser()) {
```
Comment (michaelneale) [inline blueocean-web/src/main/js/main.jsx]: heh, I just bet the anon user is localized somewhere or other hehehe

File: blueocean-web/src/main/js/main.jsx [hunk 4]
```diff
@@ -169,4 +169,4 @@ Extensions.store.getExtensions(['jenkins.main.routes', 'jenkins.main.stores'], (
 });
 
 // Enable page reload.
-require('./reload');
+// require('./reload');
```
Comment (michaelneale) [inline blueocean-web/src/main/js/main.jsx]: What is the explanation for this change?

**Comment by michaelneale** [comment]
@cliffmeyers haven't played with this change to review it, but it looks reasonable. I would like perhaps @imeredith or someone to review the changes around the security refactoring (I am not sold on the generic "Security" name for the class/refactoring, perhaps Security Context or something? but probably just me). 

If can update it, get it passing ATH and in place, then why not get it in (some good changes it looks like). Also - we will need some switch to enable/disable the create new pipeline link from opening the create dialog as for a little while still be redirecting to classic.

**Comment by cliffmeyers** [comment]
@michaelneale have any ideas for the switch? I could swap in the new button based on the presence of a query string param perhaps. `?newCreate=true` Any ideas, I'm all ears.

**Comment by michaelneale** [comment]
@cliffmeyers 

some ideas:  
* ?enableCreate=true always enables the creation flow in the URI
* if running with a /jenkins path, we know it is running in dev mode (hpi:run) so can show the new dialog (if that isn't there, could just do the link to classic)

File: blueocean-dashboard/src/main/js/PipelineRoutes.jsx [hunk 5]
```diff
@@ -75,24 +76,52 @@ function persistModalBackground() {
     root.appendChild(container);
 }
 
+function isEnteringRunDetails(prevState, nextState) {
```
Comment (michaelneale) [inline blueocean-dashboard/src/main/js/PipelineRoutes.jsx]: all these fiddly things look like great places for bugs to hide - unit testable?

**Comment by cliffmeyers** [comment]
@imeredith looking for some feedback on the Security refactor. In the mean time I'm bringing this branch up to date and making changes requested by @michaelneale

**Comment by cliffmeyers** [comment]
@michaelneale I've implemented a switch based on a "blueCreate" query param in commit 913fc97

**Comment by michaelneale** [comment]
OK. ATH passes, and I can see it opens a dialog - but it is empty (I assume this is meant to be so?) If so - then :bee: for this to move it along.

**Pull request merged**